src/include/RADOS/RADOS.hpp

```diff
@@ -0,0 +1,612 @@
+#define RADOS_UNLEASHED_HPP
+
+class CephContext;
+
+namespace RADOS {
+class Object;
+}
+namespace std {
+template<>
+struct hash<RADOS::Object>;
+}
+
+namespace RADOS {
+
+class RADOS;
+
+// Exists mostly so that repeated operations on the same object don't
+// have to pay for the string copy to construct an object_t.
+
+class Object {
+  friend RADOS;
+  friend std::hash<Object>;
+
+public:
+  Object(std::string_view s);
+  Object(std::string&& s);
+  Object(const std::string& s);
+  ~Object();
+
+  Object(const Object& o);
+  Object& operator =(const Object& o);
+
+  Object(Object&& o) = delete;
+  Object& operator =(Object&& o) = delete;
+
+  operator std::string_view() const;
+
+  friend bool operator <(const Object& lhs, const Object& rhs);
+  friend bool operator <=(const Object& lhs, const Object& rhs);
+  friend bool operator >=(const Object& lhs, const Object& rhs);
+  friend bool operator >(const Object& lhs, const Object& rhs);
+
+  friend bool operator ==(const Object& lhs, const Object& rhs);
+  friend bool operator !=(const Object& lhs, const Object& rhs);
+
+private:
+
+  static constexpr std::size_t impl_size = 4 * 8;
+  std::aligned_storage_t<impl_size> impl;
+};
+
+// Not the same as the librados::IoCtx, but it does gather together
+// some of the same metadata. Since we're likely to do multiple
+// operations in the same pool or namespace, it doesn't make sense to
+// redo a bunch of lookups and string copies.
+
+struct IOContext {
+  friend RADOS;
+
+  IOContext();
+  explicit IOContext(std::int64_t pool);
+  IOContext(std::int64_t _pool, std::string_view _ns);
+  IOContext(std::int64_t _pool, std::string&& _ns);
+  ~IOContext();
+
+  IOContext(const IOContext& rhs);
+  IOContext& operator =(const IOContext& rhs);
+
+  IOContext(IOContext&& rhs);
+  IOContext& operator =(IOContext&& rhs);
+
+  std::int64_t pool() const;
+  void pool(std::int64_t _pool);
+
+  std::string_view ns() const;
+  void ns(std::string_view _ns);
+  void ns(std::string&& _ns);
+
+  // Because /some fool/ decided to disallow optional references,
+  // you'd have to construct a string in an optional which I would then
+  // take an optional reference to. Thus a separate 'clear' method.
+  std::optional<std::string_view> key() const;
+  void key(std::string_view _key);
+  void key(std::string&& _key);
+  void clear_key();
+
+  std::optional<std::int64_t> hash() const;
+  void hash(std::int64_t _hash);
+  void clear_hash();
+
+  std::optional<std::uint64_t> read_snap() const;
+  void read_snap(std::optional<std::uint64_t> _snapid);
+
+  // I can't actually move-construct here since snapid_t is its own
+  // separate class type, not an alias.
+  std::optional<
+    std::pair<std::uint64_t,
+	      std::vector<std::uint64_t>>> write_snap_context() const;
+  void write_snap_context(std::optional<
+			  std::pair<std::uint64_t,
+			              std::vector<std::uint64_t>>> snapc);
+private:
+
+  static constexpr std::size_t impl_size = 16 * 8;
```

Comment:
They come from the static_assertions in RADOS.cc. It's the compromise between hiding the implementation data type and not having to allocate.

Really this is the thing I like least since it results in rather ugly code and removes many of the better uses of templates. One of the stated reasons for having a 'RADOS' library instead of using Objecter directly was to keep Ceph data types from leaking out into the API. It's why I want modules. Among other reasons.